File: jaros/state/replication.py
"""Log replication across nodes (EXT-002 / REQ-5).

State survives the loss of any single node because every committed transition is
mirrored to more than one log before the append is acknowledged.

``ReplicatedLog`` wraps a *primary* :class:`~jaros.state.log.TransitionLog` and
zero or more *replica* sinks (also ``TransitionLog`` instances, typically
backed by distinct files/directories). Each :meth:`append` writes to the primary
and every replica before returning, so an acknowledged append exists on all
nodes. Convergence is checked by index + checksum:
:meth:`converged_prefix` returns the longest agreed prefix length,
:meth:`has_converged` reports full agreement, and :meth:`reconcile` re-applies
missing tail entries from the most-advanced log to any lagging replica.

.. note::

   This is a **single-node, file-backed stand-in** for a true multi-node
   deployment. The "replicas" are independent on-disk logs on the same host,
   not remote nodes reached over a network. It models the *semantics* of
   replication (mirror-before-ack, index+checksum convergence, no-loss on
   single-replica loss) so the rest of the system can depend on the contract;
   a production deploy would swap the file sinks for real remote peers without
   changing this interface.
"""

from __future__ import annotations

from jaros.state.log import LogEntry, TransitionLog


# #EXT-002-REQ-5 Start
class ReplicationError(RuntimeError):
    """Raised when replication cannot satisfy its mirror-before-ack contract."""


class ReplicatedLog:
    """A transition log mirrored to one or more replica sinks before ack.

    The ``primary`` is the authoritative local log; ``replicas`` are additional
    ``TransitionLog`` sinks that receive a copy of every appended entry. An
    :meth:`append` is acknowledged only after the primary *and* every replica
    have durably stored the entry.
    """

    def __init__(
        self,
        primary: TransitionLog,
        replicas: list[TransitionLog] | None = None,
    ) -> None:
        self.primary: TransitionLog = primary
        self.replicas: list[TransitionLog] = list(replicas or [])

    def add_replica(self, replica: TransitionLog) -> None:
        """Register an additional replica sink."""
        self.replicas.append(replica)

    def ensure(self) -> None:
        """Ensure the primary and every replica log file exists."""
        self.primary.ensure()
        for replica in self.replicas:
            replica.ensure()

    def append(self, entry: LogEntry) -> None:
        """Mirror ``entry`` to the primary and every replica, then acknowledge.

        Writes the primary first, then each replica, each via the underlying
        durable (fsync-ing) :meth:`TransitionLog.append`. Only after all sinks
        have stored the entry does this return — so a committed transition is
        never present on fewer than ``1 + len(replicas)`` logs.
        """
        self.primary.append(entry)
        for replica in self.replicas:
            replica.append(entry)

    def length(self) -> int:
        """Return the primary's entry count."""
        return self.primary.length()

    def _all_logs(self) -> list[TransitionLog]:
        return [self.primary, *self.replicas]

    def converged_prefix(self) -> int:
        """Return the length of the longest prefix all logs agree on.

        Two logs agree on position ``i`` iff their entries there have the same
        ``index`` and ``checksum``. The returned value is the largest ``k`` such
        that every log has identical entries for positions ``0..k-1``.
        """
        logs = self._all_logs()
        snapshots = [list(log.read()) for log in logs]
        shortest = min((len(s) for s in snapshots), default=0)
        prefix = 0
        for i in range(shortest):
            ref = snapshots[0][i]
            if all(
                snap[i].index == ref.index and snap[i].checksum == ref.checksum
                for snap in snapshots
            ):
                prefix += 1
            else:
                break
        return prefix

    def has_converged(self) -> bool:
        """Return ``True`` iff every log holds the identical entry sequence."""
        logs = self._all_logs()
        snapshots = [list(log.read()) for log in logs]
        lengths = {len(s) for s in snapshots}
        if len(lengths) != 1:
            return False
        return self.converged_prefix() == lengths.pop()

    def reconcile(self) -> None:
        """Bring every log up to the most-advanced log's full sequence.

        Picks the log with the longest valid (checksum-passing, index-continuous)
        entry sequence as the source of truth and appends any missing tail
        entries to every other log. This models recovery of a replica that
        missed appends (e.g. it was the node that was briefly lost) — after a
        :meth:`reconcile` the logs :meth:`has_converged`.

        Raises:
            ReplicationError: if logs diverge on an already-shared position
                (a genuine conflict that simple tail-append cannot reconcile).
        """
        logs = self._all_logs()
        snapshots = [list(log.read()) for log in logs]

        # Source of truth: the longest sequence.
        source_idx = max(range(len(logs)), key=lambda i: len(snapshots[i]))
        source = snapshots[source_idx]

        # Detect genuine conflicts on shared positions before mutating anything.
        for snap in snapshots:
            for i in range(min(len(snap), len(source))):
                if (
                    snap[i].index != source[i].index
                    or snap[i].checksum != source[i].checksum
                ):
                    raise ReplicationError(
                        f"divergent entry at position {i}; cannot reconcile by "
                        "tail-append"
                    )

        for log, snap in zip(logs, snapshots):
            for entry in source[len(snap):]:
                log.append(entry)
# #EXT-002-REQ-5 End

File: src/state/replication.ts
import { checksumOf, type LogEntry, type LogEntryInput, TransitionLog } from "./log";

// #EXT-002-REQ-5 Start
/**
 * Log replication across nodes ([REQ-5]). Each appended transition is mirrored to
 * every registered replica sink *before* the append is acknowledged, so the loss
 * of any single node loses no committed transition. Replicas converge on the
 * same committed sequence via index + checksum reconciliation.
 *
 * SINGLE-NODE STAND-IN: a true deployment replicates to peer nodes over the
 * network. This implementation provides an in-process / multi-file local
 * stand-in: each {@link ReplicaSink} is an independent {@link TransitionLog}
 * (its own file). It exercises the exact replication contract — mirror before
 * ack, detect divergence by index+checksum, reconcile to a common prefix —
 * without a network, and is intended to be swapped for real remote sinks in a
 * multi-node deploy.
 */

/**
 * A replica destination. The local stand-in wraps an independent
 * {@link TransitionLog}; a real implementation would forward to a peer node.
 */
export interface ReplicaSink {
  /** Stable identifier for diagnostics / reconciliation reporting. */
  readonly id: string;
  /** Persist (mirror) an already-derived entry. Throws on failure. */
  accept(entry: LogEntry): void;
  /** Read this replica's entries in append order. */
  read(): LogEntry[];
}

/** A {@link ReplicaSink} backed by a local {@link TransitionLog} file. */
export class FileReplicaSink implements ReplicaSink {
  readonly id: string;
  private readonly log: TransitionLog;

  constructor(id: string, log: TransitionLog) {
    this.id = id;
    this.log = log.ensure();
  }

  accept(entry: LogEntry): void {
    // Mirror verbatim (preserve the primary's index/checksum) by writing the
    // exact serialized entry; the replica's own append would re-derive index.
    this.log.append({ from: entry.from, event: entry.event, to: entry.to });
  }

  read(): LogEntry[] {
    return this.log.read();
  }
}

/** Raised when a replica acknowledgement fails so the entry is not committed. */
export class ReplicationError extends Error {
  readonly replicaId: string;

  constructor(replicaId: string, cause: unknown) {
    super(`Replication to "${replicaId}" failed: ${String(cause)}`);
    this.name = "ReplicationError";
    this.replicaId = replicaId;
    Object.setPrototypeOf(this, ReplicationError.prototype);
  }
}

/**
 * Replicated log: writes to a primary {@link TransitionLog} and mirrors every
 * appended entry to all registered {@link ReplicaSink}s before acknowledging.
 */
export class ReplicatedLog {
  private readonly primary: TransitionLog;
  private readonly replicas: ReplicaSink[];

  constructor(primary: TransitionLog, replicas: ReplicaSink[] = []) {
    this.primary = primary.ensure();
    this.replicas = [...replicas];
  }

  /** Register an additional replica sink. */
  addReplica(sink: ReplicaSink): this {
    this.replicas.push(sink);
    return this;
  }

  /**
   * Append to the primary, then mirror to every replica before returning. The
   * entry is acknowledged (returned) only once it is on the primary and all
   * replicas, so a subsequent loss of any single node still leaves the entry on
   * the others. A replica failure raises {@link ReplicationError}.
   */
  append(input: LogEntryInput): LogEntry {
    const entry = this.primary.append(input);
    for (const replica of this.replicas) {
      try {
        replica.accept(entry);
      } catch (cause) {
        throw new ReplicationError(replica.id, cause);
      }
    }
    return entry;
  }

  /** All known nodes (primary first), each as an ordered entry list. */
  private nodes(): { id: string; entries: LogEntry[] }[] {
    return [
      { id: "primary", entries: this.primary.read() },
      ...this.replicas.map((r) => ({ id: r.id, entries: r.read() })),
    ];
  }

  /**
   * The longest prefix of entries on which every node agrees, by 1-indexed
   * position and per-entry checksum. This is the committed, converged sequence —
   * any committed entry (present on all nodes) is preserved here even if one
   * node later trails.
   */
  convergedPrefix(): LogEntry[] {
    const nodes = this.nodes();
    if (nodes.length === 0) {
      return [];
    }
    const minLen = Math.min(...nodes.map((n) => n.entries.length));
    const prefix: LogEntry[] = [];
    for (let i = 0; i < minLen; i++) {
      const reference = nodes[0].entries[i];
      const expectedChecksum = checksumOf(i + 1, reference.from, reference.event, reference.to);
      const allAgree = nodes.every((n) => {
        const e = n.entries[i];
        return (
          e.index === i + 1 &&
          e.from === reference.from &&
          e.event === reference.event &&
          e.to === reference.to &&
          e.checksum === expectedChecksum
        );
      });
      if (!allAgree) {
        break;
      }
      prefix.push(reference);
    }
    return prefix;
  }

  /** True when every node holds the identical committed sequence. */
  hasConverged(): boolean {
    const nodes = this.nodes();
    const target = this.convergedPrefix().length;
    return nodes.every((n) => n.entries.length === target);
  }

  /**
   * Reconcile divergent replicas toward the primary: any replica missing
   * committed entries that the primary holds is brought forward by mirroring the
   * absent suffix. This models a node rejoining after a loss and catching up so
   * all replicas converge on the same committed sequence.
   */
  reconcile(): void {
    const primaryEntries = this.primary.read();
    for (const replica of this.replicas) {
      const have = replica.read().length;
      for (let i = have; i < primaryEntries.length; i++) {
        replica.accept(primaryEntries[i]);
      }
    }
  }
}
// #EXT-002-REQ-5 End
